### ISSUE TITLE
Attenuation-law shape parameters are unfittable on single_component: dust_slope / dust_delta / dust_bump_strength have exactly-zero gradients (6 laws)

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#1808]
## Summary

On the `single_component` dust path, **every attenuation-law shape parameter has an exactly-zero gradient**. `dust_slope`, `dust_delta` and `dust_bump_strength` are declared, accepted by the grammar, sampled into the parameter dict, carried into the forward pass — and never reach the attenuation curve.

A zero gradient is not cosmetic. A fit that frees a UV slope or a 2175 Å bump on this path **cannot move it**: the sampler explores the prior, the posterior returns the prior, and it reports convergence. Nothing warns.

Both are standard SED-fitting free parameters — `dust_delta` is the Noll+2009 slope.

All measurements on `origin/main` @ `57baf39ed`, 2026-08-15.

## Measured three independent ways

**1. The value moves and the SED does not.** `dust_slope` goes −0.7 → −2.5 in the sampled parameter dict; the SED is bit-identical.

```
power_law @ default
   sampled: dust_slope = -0.7      sed sum = 1.7130496018e+46
power_law @ dust_slope=-2.5
   sampled: dust_slope = -2.5      sed sum = 1.7130496018e+46
   bit-identical SED: True
```

**2. The gradient is exactly zero** — and `dust_tau_v` on the *same build* is healthy:

```
single_component / power_law   [objective scale = 6.9287e-12]
    dust_slope           grad=+0.0000e+00   DEAD (exactly 0)
    dust_tau_v           grad=-8.3893e-12   rel=1.211e+00

single_component / noll09      [objective scale = 6.7544e-12]
    dust_bump_strength   grad=+0.0000e+00   DEAD (exactly 0)
    dust_delta           grad=+0.0000e+00   DEAD (exactly 0)
    dust_tau_v           grad=-8.5695e-12   rel=1.269e+00
```

So the build works. Only the shape parameters are dead.

**3. The control: the same parameters are live on `two_component`.**

```
two_component / noll09         [objective scale = 3.8996e-12]
    dust_bump_strength   grad=-1.5871e-14   rel=4.070e-03
    dust_delta           grad=+1.0349e-12   rel=2.654e-01
    dust_tau_bc          grad=-4.5588e-15   rel=1.169e-03
    dust_tau_diff        grad=-4.7503e-12   rel=1.218e+00
```

That is what makes this a **wiring** defect rather than a property of the laws.

> Read the gradients relative to the objective — `sum(photometry)` is ~1e-12 in these units, so an absolute 1e-12 gradient is *healthy*, not dead. Only an exactly-zero gradient is scale-free evidence. I initially misread `dust_delta`'s 1.02e-12 as near-zero before checking the scale.

## Scope: six laws

Dead on `single_component`, over a 1 Å–10⁹ Å wide-band objective:

| law | dead |
|---|---|
| `power_law` | `dust_slope` |
| `noll09` | `dust_delta`, `dust_bump_strength` |
| `salim_sbl18` | `dust_delta`, `dust_bump_strength` |
| `kriek_conroy` | `dust_delta`, `dust_bump_strength` |
| `narayanan_z` | `dust_delta`, `dust_bump_strength` |
| `tea` | `dust_delta` |

`calzetti`, `smc`, `leitherer02`, `li08`, `reddy15` have no shape parameters and are unaffected. `two_component/{calzetti,noll09,kriek_conroy}` are clean.

## This silently collapses the attenuation menu

With bump and slope frozen, four law names reduce to two curves:

- `leitherer02` ≡ `noll09` ≡ `salim_sbl18` — both modifications default to zero, so each reduces to the unmodified Calzetti+L02 base.
- `power_law` ≡ `vw07_diff` — `power_law`'s slope is stuck at its default −0.7, and `vw07_diff` is a fixed −0.7 power law.

A user selecting `salim_sbl18` on `single_component` receives the Calzetti+L02 curve, with no warning and no way to change it. That is the #1738 fail-open shape, at the parameter level.

## Suggested treatment

The `two_component` path already threads these correctly, so the fix is presumably to route the same parameters through the `single_component` lowering rather than new physics — but I have not traced where they are dropped, so treat that as a hypothesis rather than a diagnosis.

Whatever the fix, the guard belongs with it: a free parameter is a promise that a fit can constrain it, and an exactly-zero gradient breaks that promise silently.

## Already pinned

Pinned by `tests/contract/test_registry_choice_is_distinct.py` in #1807, in `KNOWN_DEAD_PARAMS`. That ledger asserts the defect is **still present**, so fixing it turns the entry red with "remove this entry" and the parameter is then protected by the general census from that point on.

Related: #1738 (the epic this is an instance of), #1093 (`dust_qpah` inert on astrodust — same shape, still reproducing on main), #1684, #1461.

**Comment by suchethac** [comment]
## Root cause located — and it is this epic's own seam, not a missing kwarg

`src/tengri/components/dust/component.py`. `DustAttenuationSEDComponent` is what `dust={'type': 'single_component'}` resolves to (`_REGISTRY["single_component"] = DustAttenuationSEDComponent`, `component.py:366`).

**It never passes law parameters to the curve.** Six sites resolve the law and evaluate it on the wavelength grid alone — `component.py:206-207, 248-249, 293, 304, 324, 330, 350`:

```python
law_fn = resolve_dust_law(self.config.law)
k = law_fn(state.wave)          # <- no **law_params
```

Compare `two_component.py:431`, which resolves them explicitly before building the curve:

```python
bc_law_params, diff_law_params = resolve_bc_diff_law_params(...)
```

`resolve_bc_diff_law_params` (`_apply.py`) is the function that maps the flat `dust_slope` / `dust_delta` / `dust_bump_strength` / `dust_Rv` onto the law-function keywords via `_TWO_COMPONENT_LAW_PARAMS`. `single_component` has no equivalent. That asymmetry is the whole defect, and it explains the measurement exactly: `apply()` reads `tau_v = jnp.asarray(params["dust_tau_v"])` per call (`component.py:251`), which is why `dust_tau_v` is live while every shape parameter is exactly zero.

### Why this is not a one-line fix

`k(λ)` is not computed per call. It is built in `precompute()` and cached on the component state (`component.py:201-208`):

```python
return DustAttenuationSEDComponentState(name=self.name, k_lambda=k)
```

and `apply()` prefers the cached array (`component.py:243-244`). **A precomputed `k(λ)` structurally cannot accommodate a free shape parameter** — the curve is frozen at build time, before any parameter value exists. Adding `**law_params` at the six call sites without addressing the cache would fix the lazy fallback path and leave the primary path frozen.

So this is an instance of exactly what #1738 is about: a precomputed state that silently freezes something a caller is still allowed to vary. The `dust_tau_v` / `k(λ)` split is the tell — the parameter that stayed outside the cached state is the one that still works.

### That makes the fix a design decision

Roughly: `precompute()` may only cache `k(λ)` when the selected law's shape parameters are all fixed; when any is free, `k` must be computed per call from `params`. `_law_shape_params` (already used by `parse_groups` to scope the group wildcard, per the note on `dust_bump_strength` in `_params.py`) knows which parameters a given law reads, so the information needed to make that decision is already available.

There is a performance consequence — a per-call curve evaluation on the hot path — which is why I am flagging this rather than pushing a patch: it is a speed/correctness trade for the maintainer, not a mechanical edit.

### Not a scoping artifact

Checked, because it would have been the obvious false positive: `ATTENUATION_TWO_COMPONENT_ONLY` (`_params.py:430`) is `frozenset({"dust_tau_bc", "dust_tau_diff"})` — the shape parameters are explicitly **not** excluded from `single_component`. And `parse_groups` narrows the wildcard to the parameters the selected law actually names, so `dust_slope` appearing as free under `power_law` is the designed behavior. These parameters are meant to be live here.

**Comment by suchethac** [comment]
**Sequencing constraint, from a parallel session working the #1738 nebular band-integration thread.**

If the nebular band-integration fix is extended to `single_component`, the natural implementation computes `sed_neb * exp(-tau_v * k)` inside the component and projects it through the same seam. That must **not** read the cached `k` described above.

`apply()` prefers the cached array (`component.py:243-244`) precisely because it was frozen at precompute. A nebular term that reads the same cached `k` inherits the freeze — and then fixing this issue would move the *stellar* term onto the live curve while leaving the *nebular* term on the stale one. Two screens disagreeing inside one model is worse than today's uniform staleness.

So either this issue lands first, or that fix must resolve `k` the same way `apply()` will resolve it afterwards, in a way that cannot diverge.

Recording it here so the constraint outlives the conversation it came from. (For the avoidance of doubt, these are two independent defects with different fixes: this one is a shape-parameter freedom defect with exactly-zero gradients that affects the exact path; that one is a band-integration bias confined to `predict_via_precomp`, where gradients are fine.)

**Comment by suchethac** [comment]
## I implemented the fix, measured it, and reverted it. Here is why.

The obvious repair works for the gradients and **breaks the physics**, and the reason is worth recording before anyone else tries it.

### The implementation

Two changes to `components/dust/component.py`:

1. Resolve one bound curve at the top of `apply()` — `law_fn` with its shape parameters read from `params` — and use it at all six `resolve_dust_law` sites (the SED screen, the filter LUT, its finite-difference slope, the sub-band quadrature, the rest band, the spectroscopy pixels). One curve rather than six call sites, so the paths cannot diverge.
2. `precompute()` caches `k(λ)` **only** when the selected law reads no shape parameters (`_law_shape_params(law)` is empty). `calzetti`, `smc`, `lmc` and friends keep the cache and the fast path unchanged — they are the default and the common case. Laws that read a shape parameter get an unprimed state, and `apply()` builds the curve from `params`.

### It fixes the reported defect

| build | parameter | before | after |
|---|---|---|---|
| `single_component/power_law` | `dust_slope` | **exactly 0.0** | rel **0.173** |
| `single_component/noll09` | `dust_delta` | **exactly 0.0** | rel **0.152** |
| `single_component/noll09` | `dust_bump_strength` | **exactly 0.0** | rel 2.19e-03 |
| `two_component/noll09` *(control)* | all four | 0.265 / 4.07e-03 / 1.17e-03 / 1.218 | **identical to every digit** |

All seven entries in the census's dead-parameter ledger self-invalidated. The control path is untouched.

### And it silently changes three published attenuation laws

The screen previously called each law with **no arguments**, so every law used its own *function signature* defaults. Those are the published values:

```
kriek_conroy   dust_bump_strength = 1.0     (KC13 as published)
salim          dust_bump_strength = 0.0
narayanan_z    dust_delta = -0.2            (Narayanan+2018 as published)
```

The parameter spec declares **one shared** `dust_delta` and `dust_bump_strength`, both `Fixed(0.0)`. Passing them unconditionally overrides each law's published default with the generic one — so Narayanan+2018's δ = −0.2 becomes 0.0, and Kriek & Conroy's bump amplitude becomes bump-free.

Measured: `kriek_conroy`, `narayanan_z` and `salim` become **bit-identical** to each other after the change. They are three distinct published laws (ApJL 775 L16; ApJ 869 70, which even takes its own `redshift` argument; ApJ 859 11), and before the change they were genuinely distinct.

So the fix trades an unfittable parameter for three silently-collapsed laws. That is the same defect class, moved.

### Why the obvious repair to the repair does not work either

The natural patch is "pass the parameter only when it differs from its declared default, so an unset parameter leaves the law's own default alone". That cannot be written here: the comparison is against a **traced** value inside `apply()`, and branching on it raises `ConcretizationTypeError`. The decision has to be made at build time, where the spec knows which parameters the user set or freed — and the component does not have the spec.

### So the decision is genuinely upstream of the component

Roughly: the build has to tell the component which shape parameters are *live* (user-set or free) versus merely defaulted, so the component can pass the former and let the law's own default stand for the latter. That is a config/plumbing change through `component_factory`, not a change inside the screen, and it needs someone to rule on what `dust_delta = 0.0` should mean for a law whose published default is −0.2 — the shared parameter and the per-law default genuinely disagree, and only one of them can win.

There is also a smaller question sitting underneath: `two_component` **already** passes the spec values, so `kriek_conroy` on that path is already bump-free while on `single_component` it was KC13-as-published. The two dust paths disagree about the same law today. Whatever is decided should make them agree.

I have left this issue open and unfixed rather than ship either half. The census in #1807 continues to pin the dead parameters, so a fix will turn those entries red and force them to be removed with it.

**Comment by suchethac** [comment]
**Fixed in #1807.** The design question in my previous comment has an answer: **provenance**, resolved at build time.

`spec._group_provenance` already distinguishes a request from a default:

| tag | meaning | pass to the law? |
|---|---|---|
| `user_prior` / `user_fixed` / `user_free` | the caller named this parameter | **yes** |
| `wildcard_free` | `all_params: FREE` freed it | **yes** |
| `registry_default` | nobody touched it | **no** — the law's own default stands |
| `wildcard_fixed` | `all_params: FIXED` pinned it | **no** — not a request either |

`SEDModel._build_component_chain` resolves the selected law's shape parameters against that and threads the requested ones into `DustAttenuationSEDComponentConfig.live_shape_params`. `apply()` passes exactly those.

This is why it could not be done inside the component: *"pass it only if it differs from the default"* is a comparison against a **traced** value. Deciding once at build time makes it a static Python branch, and the component sees only a plain set of names.

### It fixes the defect

| build | parameter | before | after |
|---|---|---|---|
| `single_component/power_law` | `dust_slope` | **exactly 0.0** | rel **0.173** |
| `single_component/noll09` | `dust_delta` | **exactly 0.0** | rel **0.152** |
| `single_component/noll09` | `dust_bump_strength` | **exactly 0.0** | rel 2.19e-03 |
| `single_component/*` | `dust_tau_v` | 1.43 | 1.43 (unchanged) |
| `two_component/noll09` *(control)* | all four | 0.265 / 4.07e-03 / 1.17e-03 / 1.218 | **identical to every digit** |

### And it does not repeat the mistake in my previous comment

`kriek_conroy`, `narayanan_z` and `salim` **stay distinct at default settings**. Their published values — bump=1.0, δ=−0.2, bump=0.0 — survive, because nobody asked for anything else. That assertion is now a test, so the naive fix cannot be reintroduced silently.

`precompute()` still caches `k(λ)` whenever nothing was requested, which is the default and the common case (`calzetti` among it), so the fast path is unchanged for any model that does not steer its curve.

### One structural change worth noting

The six places `apply()` evaluated the curve — the SED screen, the filter LUT, that LUT's finite-difference slope, the sub-band quadrature, the rest band, the spectroscopy pixels — are now one bound curve resolved at the top. Not tidying: they must all use the same curve with the same parameters, and six independent call sites are six chances for one to keep the old behaviour and put two different screens inside one model.

### Still open underneath this

`two_component` has always passed the spec values, so on that path `kriek_conroy` was already bump-free while `single_component` used KC13-as-published. **The two dust paths still disagree about the same law at default.** This change does not resolve that — it makes `single_component` honour requests without discarding published defaults, but the question of which default `two_component` should use is untouched and worth its own issue if you want them aligned.

Tests: `tests/regression/bug/test_single_component_law_shape_params.py`. 7949 passed / 138 skipped / 9 xfailed across `tests/contract/`, `tests/components/dust/`, `tests/regression/bug/` and `tests/regression/precision/`.

The census's `KNOWN_DEAD_PARAMS` ledger self-invalidated on this and is now empty; the two dust classes it recorded as defects moved to `DECLARED_COINCIDENT`, since their separators run on `single_component` now.

**Comment by suchethac** [comment]
Related, found while measuring the consolidation this PR's fix pointed at: **#1833**.

The approach #1808 implemented, measured and *rejected* for `single_component` — pass the spec's shared `dust_delta`/`dust_bump_strength` unconditionally, which "collapsed three distinct published laws onto one curve" — is what `two_component` has always done. Same three laws (`kriek_conroy`, `narayanan_z`, `tea`), measured at up to **128%** on the SED.

So #1808 fixed the path with no users and left the path all ten recipes build. The provenance rule established here is the fix; it needs a second caller.
